recursive-feature-elimination/script.whizzml

```diff
@@ -0,0 +1,291 @@
+  ;;
+  ;; Selects the field importances in an ensemble
+  ;;
+  ;; Inputs:
+  ;;   ensemble: (ensemble) Ensemble information
+  ;;
+  ;; Outputs: (map) Importances by field name
+
+  (define (get-importances ensemble)
+    (let (dist (ensemble "distributions" false)
+          imps (map (lambda (x) (x "importance" false)) dist)
+          m-num (count imps))
+      (iterate (acc {} m-imps imps)
+        (iterate (m-acc acc imp m-imps)
+          (assoc+ m-acc
+                  (imp 0)
+                  (/ (imp 1) m-num))))))
+
+
+  ;; assoc+
+  ;;
+  ;; Adds the value in the key-value pair to the one existing in the accumulator
+  ;;
+  ;; Input:
+  ;;   acc: (map) accumulator map
+  ;;   key: (string) key to accumulate
+  ;;   value: (number) value to be added
+  (define (assoc+ acc key value)
+    (assoc acc key (+ value (acc key 0))))
+
+
+
+  ;; recursive-feature-elimination
+  ;;
+  ;; Applies recursive feature elimination algorithm and returns a map with
+  ;; the selected fields names, selected fields ids, and information about
+  ;; each iteration
+  ;;
+  ;; Inputs:
+  ;;   dataset-id: (dataset-id) Dataset id (for training)
+  ;;   nfeatures: (number) Number of features that we want to obtain
+  ;;   objective-id: (string) Objective field id
+  ;;   test-dataset-id: (dataset-id) id of the dataset used to
+  ;;                     evaluate the model
+  ;;   evaluation-metric: (string) Metric  used in the evaluation
+  ;; Output: (map) Map with selected field names, selected field ids, and 
+  ;;         iterations evaluation info (only if test-dataset-id is present)
+  (define (recursive-feature-elimination dataset-id
+                                         nfeatures
+                                         objective-id
+                                         test-dataset-id
+                                         evaluation-metric)
+    (let (do-evaluation (not (empty? test-dataset-id))
+          obj-id (get-objective dataset-id objective-id)
+          fields (resource-fields dataset-id)
+          input-ids (default-inputs dataset-id obj-id))
+      (loop (selected input-ids
+             evaluations []) 
+          ;; Loop until we obtain the desired number of features
+          (if (<= (count selected) nfeatures)
+              (response selected 
+                        (feature-names fields selected) 
+                        evaluations 
+                        do-evaluation)
+              (let (_ (log-info "------- New iteration -------")
+                    _ (log-info "Creating random forest with " 
+                                (count selected) 
+                                " fields")
+                    ;; Create random forest and obtain its importances
+                    rf-id (random-forest dataset-id obj-id 0 selected)
+                    imps (get-importances (fetch rf-id))
+                    ;; Obtaining minimum field importance
+                    min-imp (minimum-importance imps)
```

Comment:
I don't see that you use the min-imp for anything other than removing the least important field... couldn't you just look for the last field as ordered by importance and remove that?

whizzml> (define c {"a" 2 "b" 1})
{"a" 2 "b" 1}
whizzml> (min-key (lambda (x) (c x)) (keys c))

Reply:
Didn't know this procedure, very interesting